lib/elixir/src/elixir_clauses.erl

```diff
@@ -40,8 +40,28 @@ guard({'when', Meta, [Left, Right]}, E) ->
   {ELeft, EL}  = guard(Left, E),
   {ERight, ER} = guard(Right, EL),
   {{'when', Meta, [ELeft, ERight]}, ER};
-guard(Other, E) ->
-  elixir_expand:expand(Other, E).
+guard(Guard, E) ->
+  {EGuard, EG} = elixir_expand:expand(Guard, E),
+  ok = warn_zero_length_guard(EGuard, EG),
+  {EGuard, EG}.
+
+warn_zero_length_guard(
+  {
+    {'.', _, [erlang, '==']},
+    Meta,
+    [{{'.', _, [erlang, length]}, _, [Arg]}, 0]
+  },
+  E
+) ->
+  ArgS = 'Elixir.Macro':to_string(Arg),
+  MessageFmt = "\"length(~ts) == 0\" is discouraged since it has to traverse the whole list to check if it is empty or not. Prefer \"~ts == []\" instead",
+  elixir_errors:warn(?line(Meta), ?key(E, file), io_lib:format(MessageFmt, [ArgS, ArgS])),
+  ok;
```

Comment:
I would not return anything different than the return value of `elixir_errors:warn/3` here since we don't use the return value, and I would not do `ok = ...` when calling this above.

Reply:
Well, `elixir_errors:warn` will return `ok` anyways, so you're right. About matching it to `ok`, any reason for it? Is the reason one less match to be computed?

Reply:
We can match on `ok` if you prefer, it doesn't have a lot of value since we're interested in the side effect of the `warn_zero_length_guard` call, not in its result but it doesn't hurt either.

Reply:
No need for that, I removed it.